lightning/src/ln/chan_utils.rs

```diff
@@ -550,6 +556,415 @@ pub fn build_htlc_transaction(prev_hash: &Txid, feerate_per_kw: u32, contest_del
+	pub htlc_sigs: Vec<Signature>,
+}
+
+impl PartialEq for HolderCommitmentTransactionInfo {
+	// We dont care whether we are signed in equality comparison
+	fn eq(&self, o: &Self) -> bool {
+		self.info == o.info
+	}
+}
+
+impl_writeable!(HolderCommitmentTransactionInfo, 0, {
+	info, counterparty_sig, htlc_sigs
+});
+
+impl HolderCommitmentTransactionInfo {
+	#[cfg(test)]
+	pub fn dummy() -> Self {
+		let dummy_key = PublicKey::from_secret_key(&Secp256k1::new(), &SecretKey::from_slice(&[42; 32]).unwrap());
+		let dummy_sig = Secp256k1::new().sign(&secp256k1::Message::from_slice(&[42; 32]).unwrap(), &SecretKey::from_slice(&[42; 32]).unwrap());
+
+		HolderCommitmentTransactionInfo {
+			info: CommitmentTransactionInfo {
+				commitment_number: 0,
+				to_broadcaster_value_sat: 0,
+				to_countersignatory_value_sat: 0,
+				feerate_per_kw: 0,
+				htlcs: Vec::new(),
+				keys: TxCreationKeys {
+					per_commitment_point: dummy_key.clone(),
+					revocation_key: dummy_key.clone(),
+					broadcaster_htlc_key: dummy_key.clone(),
+					countersignatory_htlc_key: dummy_key.clone(),
+					broadcaster_delayed_payment_key: dummy_key.clone(),
+				}
+			},
+			counterparty_sig: dummy_sig,
+			htlc_sigs: Vec::new()
+		}
+	}
+	pub(crate) fn txid<T: secp256k1::Signing + secp256k1::Verification>(&self, channel_static_info: &ChannelStaticInfo, secp_ctx: &Secp256k1<T>) -> Txid {
+		self.info.txid(true, channel_static_info, secp_ctx)
+	}
+}
+
+impl HolderCommitmentTransactionInfo {
+	pub(crate) fn to_holder_commitment_tx<T: secp256k1::Signing + secp256k1::Verification>(&self, channel_static_info: &ChannelStaticInfo, secp_ctx: &Secp256k1<T>) -> HolderCommitmentTransaction {
+		let opt_sigs: Vec<Option<Signature>> = self.htlc_sigs.iter().map(|s| Some(s.clone())).collect();
+		let mut htlcs = self.info.htlcs.clone();
+		let htlcs_with_sig = htlcs.drain(..).zip(opt_sigs).collect();
+		self.info.to_holder_commitment_tx(self.counterparty_sig, htlcs_with_sig, channel_static_info, secp_ctx)
+	}
+}
+
+/// This class tracks the information needed to build a commitment transaction and to actually
+/// build and sign.  It is used for holder transactions that we sign only when needed
+/// and for transactions we sign for the counterparty.
+///
+/// This class can be used inside a signer implementation to generate a signature given the relevant
+/// secret key.
+#[derive(PartialEq)]
+#[derive(Clone)]
+pub struct CommitmentTransactionInfo {
+	/// The backwards-counting commitment number
+	pub commitment_number: u64,
+	/// The value to be sent to the broadcaster
+	pub to_broadcaster_value_sat: u64,
+	/// The value to be sent to the counterparty
+	pub to_countersignatory_value_sat: u64,
+	/// The feerate paid per 1000-weight-unit in this commitment transaction.
+	pub feerate_per_kw: u32,
+	/// The HTLCs which were included in this commitment transaction in output order.
+	pub htlcs: Vec<HTLCOutputInCommitment>,
+	pub(crate) keys: TxCreationKeys,
```

Comment:
Can you add a comment to indicate it's covering both broadcaster/countersignatory keys.

Reply:
wouldn't that duplicate the `TxCreationKeys` struct docs?

Reply:
IMO, I prefer a bit of comment overlap to avoid reviewer grep'ing the struct but that's up to you.

Reply:
OK, adding something to this effect